File: backend/src/utils/telegramService.js
import { env } from '../config/env.js';

/**
 * Send a notification message to Telegram via Bot API
 * @param {string} text Markdown formatted message
 */
export async function sendTelegramNotification(text) {
  const { telegramBotToken, telegramChatId } = env;

  if (!telegramBotToken || !telegramChatId) {
    console.log('[Telegram Alert Skipped] TELEGRAM_BOT_TOKEN or TELEGRAM_CHAT_ID not configured in env.');
    return false;
  }

  try {
    const url = `https://api.telegram.org/bot${telegramBotToken}/sendMessage`;
    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        chat_id: telegramChatId,
        text,
        parse_mode: 'Markdown'
      })
    });

    const data = await response.json();
    if (!data.ok) {
      console.error('[Telegram API Error]:', data.description);
      return false;
    }

    return true;
  } catch (error) {
    console.error('[Telegram Notification Failed]:', error.message);
    return false;
  }
}

File: backend/app/services/telegram_service.py
import os
import json
import urllib.request
import logging

logger = logging.getLogger(__name__)

TELEGRAM_BOT_TOKEN = os.getenv("TELEGRAM_BOT_TOKEN", "8992926370:AAGogt3wrygo2YEocPd3rIW6Uxzto0LCdcM")
TELEGRAM_CHAT_ID = os.getenv("TELEGRAM_CHAT_ID", "1352837172")

def send_telegram_notification(text: str) -> bool:
    """
    Sends instant Markdown formatted alerts to your Telegram chat via Bot API.
    """
    token = os.getenv("TELEGRAM_BOT_TOKEN", TELEGRAM_BOT_TOKEN)
    chat_id = os.getenv("TELEGRAM_CHAT_ID", TELEGRAM_CHAT_ID)

    if not token or not chat_id:
        logger.info("[Telegram Alert Skipped] Token or Chat ID not configured.")
        return False

    url = f"https://api.telegram.org/bot{token}/sendMessage"
    payload = {
        "chat_id": chat_id,
        "text": text,
        "parse_mode": "Markdown"
    }

    try:
        data = json.dumps(payload).encode("utf-8")
        req = urllib.request.Request(
            url,
            data=data,
            headers={"Content-Type": "application/json"}
        )
        with urllib.request.urlopen(req, timeout=5) as response:
            res_data = json.loads(response.read().decode("utf-8"))
            if res_data.get("ok"):
                logger.info("✅ Telegram notification sent successfully!")
                return True
            else:
                logger.error(f"❌ Telegram API Error: {res_data.get('description')}")
                return False
    except Exception as e:
        logger.error(f"❌ Telegram notification failed: {e}")
        return False
